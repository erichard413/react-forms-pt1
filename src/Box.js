import React from 'react';
import './Box.css';

const Box = ({width, height, color, idx, handleXClick}) => {
return (
    <>
        <div className="Box" style={{
            backgroundColor: color,
            width: `${width}px`,
            height: `${height}px`
        }}>
    </div>
    <span role="delete" className="redX" onClick={()=>handleXClick(idx)}>{`\u{274C}`}</span>
    </>

    )
}

export default Box;