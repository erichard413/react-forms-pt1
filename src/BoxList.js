import { BLOCK_TYPES } from '@babel/types';
import React, {useState} from 'react';
import Box from './Box';
import {v4 as uuidv4} from 'uuid';
import NewBoxForm from './NewBoxForm';

// BoxList - Place your state that contains all of the boxes here. This component should render all of the Box components along with the NewBoxForm component

// You should be able to specify the Box’s width, height, and background color. When the form is submitted, clear the input values.

// When each Box component is displayed, add a button with the text of of “X” next to each Box. When this button is clicked, remove that specific box. This will require you to pass a function down as props - the button should not be a separate component, it should be included in the Box component.

const BoxList = () => {
    const INITIAL_STATE = [
    ]
    const [boxes, setBoxes] = useState(INITIAL_STATE)
    const removeBox = (idx) => {
        setBoxes(boxes => {
            const boxesCopy = [...boxes];
            return boxesCopy.filter(b => b.idx !== idx);
        }
        )
    }

    const addBox = (width, height, color) => {
        setBoxes(boxes => {
            const boxesCopy = [...boxes, {
                width: width,
                height: height,
                color: color,
                idx: uuidv4()
            }]
            return boxesCopy;
        })
    }

    return (
        <>
            <NewBoxForm addBox={addBox}/>
            <div className="BoxList">
            {boxes.map(({color, height, width, idx}) => (
            <Box color={color} idx={idx} key={idx} height={height} width={width} handleXClick={removeBox} /> 
         ))}
        </div>
        </>


    )
}

export default BoxList