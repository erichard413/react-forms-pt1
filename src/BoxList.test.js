import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import BoxList from './BoxList';
import Box from './Box';

it('should render', ()=> {
    render(<BoxList />);
})

it('should match snapshot', ()=> {
    const {asFragment} = render(<BoxList />);
    expect(asFragment()).toMatchSnapshot();
});

it('should start with no boxes', ()=> {
    const result = render(<BoxList />);
    const box = result.container.querySelector('.Box');
    expect(box).not.toBeInTheDocument();
})

it('should add new box', ()=> {
    const result = render(<BoxList />);
    const {getByLabelText, queryByText} = result;
    const width = getByLabelText('Width');
    const height = getByLabelText('Height');
    const color = getByLabelText('Color');
    const btn = queryByText('Create Box');
    fireEvent.change(width, {target: {value: '200'}});
    fireEvent.change(height, {target: {value: '200'}});
    fireEvent.change(color, {target: {value: 'teal'}});
    fireEvent.click(btn);
    const box = result.container.querySelector('.Box');
    expect(box).toBeInTheDocument();
})

it('should delete box', ()=> {
    const result = render(<BoxList />);
    const {getByLabelText, queryByText, getByRole} = result;
    const width = getByLabelText('Width');
    const height = getByLabelText('Height');
    const color = getByLabelText('Color');
    const btn = queryByText('Create Box');
    
    fireEvent.change(width, {target: {value: '200'}});
    fireEvent.change(height, {target: {value: '200'}});
    fireEvent.change(color, {target: {value: 'teal'}});
    fireEvent.click(btn);
    const delBtn = getByRole('delete');

    const box = result.container.querySelector('.Box');
    expect(box).toBeInTheDocument();
    fireEvent.click(delBtn);
    expect(box).not.toBeInTheDocument();
})