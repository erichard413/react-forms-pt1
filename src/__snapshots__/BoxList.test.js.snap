// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should match snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      for="width"
    >
      Width
    </label>
    <input
      id="width"
      name="width"
      placeholder="width"
      type="text"
      value=""
    />
    <label
      for="height"
    >
      Height
    </label>
    <input
      id="height"
      name="height"
      placeholder="height"
      type="text"
      value=""
    />
    <label
      for="color"
    >
      Color
    </label>
    <input
      id="color"
      name="color"
      placeholder="color"
      type="text"
      value=""
    />
    <button>
      Create Box
    </button>
  </form>
  <div
    class="BoxList"
  />
</DocumentFragment>
`;
